crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,152 @@
+use crate::component::{Component, ComponentTicks};
+use bevy_reflect::Reflect;
+use std::ops::{Deref, DerefMut};
+
+pub trait ChangeDetectable {
+    /// Returns true if (and only if) this value been added since the last execution of this
+    /// system.
+    fn is_added(&self) -> bool;
+
+    /// Returns true if (and only if) this value been changed since the last execution of this
+    /// system.
+    fn is_changed(&self) -> bool;
+
+    /// Manually flags this value as having been changed. This normally isn't
+    /// required because accessing this pointer mutably automatically flags this
+    /// value as "changed".
+    ///
+    /// **Note**: This operation is irreversible.
+    fn set_changed(&mut self);
```

Comment:
Do we want a `set_unchanged` method as well? It seems like a nice trivial implementation that would be *very* tricky to work around as a user in case you need that behavior and I'd rather not have people using dumb interior mutability tricks to bypass change detection.

Reply:
The problem with `set_unchanged` is that it erases history from a "global" context. This is different from setting something to changed, which wont prevent previous changes from being picked up.

A good example of this is something like Transform. If we let users "set_unchanged" on that to ignore a local change in their app logic, changes made elsewhere (ex a physics plugin) won't get picked up by transform_propagate_system.

Reply:
Right, and that's likely to cause painful and insidious bugs down the road. 

Okay, I can get behind that logic.